src/nodes.rs

```diff
@@ -0,0 +1,93 @@
+use crate::geoip_cache::GeoIPCache;
+use ziggurat_core_geoip::geoip::GeoInfo;
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+use spectre::graph::Graph;
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+pub struct Node {
+    pub ip: String,
+    pub betweenness: f64,
+    pub closeness: f64,
+    pub column_position: u32,
+    pub column_size: u32,
+    pub connections: Vec<usize>,
+    pub geolocation: Option<GeoInfo>,
+}
+
+fn hash_geo_location(latitude: f64, longitude: f64) -> String {
+    // make unique every 0.2, so multiply by 5, convert to integer
+    let ilatitude: i32 = (latitude * 5.0).floor() as i32;
+    let ilongitude: i32 = (longitude * 5.0).floor() as i32;
+    format!("{ilatitude}:{ilongitude}")
+}
+
+// essentially, we sort the nodes into groups at (nearly) same geo-location
+// we use 0.2 degrees for epsilon in both axes.
+// hash gets created from a string created by two numbers
+// increment each time the same location is found
+pub fn set_column_positions(nodes: &mut Vec<Node>) -> HashMap<String, u32> {
+    let mut column_stats: HashMap<String, u32> = HashMap::new();
```

Comment:
Stating the type here (probably) isn't needed.

Reply:
done